code/game/machinery/hologram.dm

```diff
@@ -68,204 +95,304 @@ var/list/holopads = list()
+			for(var/I in holopads)
+				var/area/A = get_area(I)
+				if(A)
+					LAZYADD(callnames[A], I)
+			callnames -= get_area(src)
+
+			var/result = input(usr, "Choose an area to call", "Holocall") as null|anything in callnames
+
+			if(qdeleted(usr) || !result || outgoing_call)
+				return
+
+			if(usr.loc == loc)
+				temp = "Dialing...<br>"
+				temp += "<A href='?src=\ref[src];mainmenu=1'>Main Menu</A>"
+				new /datum/holocall(usr, src, callnames[result])
+
+	else if(href_list["connectcall"])
+		var/datum/holocall/call_to_connect = locate(href_list["connectcall"])
+		if(!qdeleted(call_to_connect))
+			call_to_connect.Answer(src)
+		temp = ""
+
+	else if(href_list["disconnectcall"])
+		var/datum/holocall/call_to_disconnect = locate(href_list["disconnectcall"])
+		if(!qdeleted(call_to_disconnect))
+			call_to_disconnect.Disconnect(src)
+		temp = ""
+
+	else if(href_list["mainmenu"])
+		temp = ""
+		if(outgoing_call)
+			outgoing_call.Disconnect()
+
+	updateDialog()
+
+//do not allow AIs to answer calls or people will use it to meta the AI sattelite
 /obj/machinery/hologram/holopad/attack_ai(mob/living/silicon/ai/user)
-	if(!istype(user))
+	if (!istype(user))
 		return
 	/*There are pretty much only three ways to interact here.
 	I don't need to check for client since they're clicking on an object.
 	This may change in the future but for now will suffice.*/
 	if(user.eyeobj.loc != src.loc)//Set client eye on the object if it's not already.
 		user.eyeobj.setLoc(get_turf(src))
-	else if(!hologram)//If there is no hologram, possibly make one.
-		activate_holo(user, 0)
-	else if(master == user)//If there is a hologram, remove it. But only if the user is the master. Otherwise do nothing.
-		clear_holo()
-	return
+	else if(!masters[user])//If there is no hologram, possibly make one.
+		activate_holo(user, 1)
+	else//If there is a hologram, remove it.
+		clear_holo(user)
+
+/obj/machinery/hologram/holopad/process()
+	for(var/I in masters)
+		var/mob/living/master = I
+		var/mob/living/silicon/ai/AI = master
+		if(!istype(AI))
+			AI = null
+
+		if(!qdeleted(master) && !master.incapacitated() && master.client && (!AI || AI.eyeobj))//If there is an AI attached, it's not incapacitated, it has a client, and the client eye is centered on the projector.
+			if(!(stat & NOPOWER))//If the  machine has power.
+				if(AI)	//ais are range based
+					if(HOLOPAD_MODE == 0)
+						if(get_dist(AI.eyeobj, src) <= holo_range)
+							return 1
+					else
+						var/area/holo_area = get_area(src)
+						var/area/eye_area = get_area(AI.eyeobj)
+
+						if(eye_area != holo_area)
+							clear_holo(master)
+							return 1
+
+					var/obj/machinery/hologram/holopad/pad_close = get_closest_atom(/obj/machinery/hologram/holopad, holopads, AI.eyeobj)
+					if(get_dist(pad_close, AI.eyeobj) <= holo_range)
+						pad_close.unset_holo(AI)
+						pad_close.activate_holo(AI, 1)
+				else
+					continue
+		clear_holo(master)//If not, we want to get rid of the hologram.

```

Comment:
The only means of avoiding clearing the hologram is via returning in the above loop.